source/SIL.AppBuilder.Portal/src/routes/(unauthenticated)/api/projects/[id=idNumber]/token/+server.ts

```diff
@@ -0,0 +1,251 @@
+  if (readOnly === null) {
+    const authors = await prisma.authors.findMany({
+      where: {
+        ProjectId: projectId
+      },
+      select: {
+        UserId: true
+      }
+    });
+    if (authors.find((a) => a.UserId === user[0].Id)) {
+      // TODO: Kalaam now wants authors to be able to update at any time.  In the future, we can add a setting on the author to whether they are a restricted author or not. I don't have time to add the UI at the moment.
+      //readOnly = !author.CanUpdate;
+      readOnly = false;
+    }
+  }
+
+  if (readOnly === null) {
+    return new Response(
+      JSON.stringify({
+        errors: [
+          {
+            title: `Project id=${projectId}, user='${user[0].Name}' with email='${user[0].Email}' does not have permission to access`
+          }
+        ]
+      }),
+      {
+        status: 403
+      }
+    );
+  }
+
+  if (tokenUse && tokenUse === TOKEN_USE_UPLOAD && readOnly) {
+    return new Response(
+      JSON.stringify({
+        errors: [
+          {
+            title: `Project id=${projectId}, user='${user[0].Name}' with email='${user[0].Email}' does not have permission to Upload`
+          }
+        ]
+      }),
+      {
+        status: 403
+      }
+    );
+  }
+
+  const tokenResult = await BuildEngine.Requests.getProjectAccessToken(
+    { type: 'query', organizationId: project.OrganizationId },
+    project.WorkflowProjectId,
+    {
+      name: project.Owner.ExternalId ?? '',
+      ReadOnly: readOnly
+    }
+  );
+
+  if (!tokenResult || tokenResult.responseType === 'error') {
+    return new Response(
+      JSON.stringify({
+        errors: [{ title: `Project id=${projectId}: GetProjectToken returned null` }]
+      }),
+      {
+        status: 400
+      }
+    );
+  }
+  if (tokenResult.SecretAccessKey == null) {
+    return new Response(
+      JSON.stringify({
+        errors: [{ title: `Project id=${projectId}: Token.SecretAccessKey is null` }]
+      }),
+      {
+        status: 400
+      }
+    );
+  }
+  const projectToken = {
+    type: 'project-tokens',
+    attributes: {
+      id: projectId,
+      url: project.WorkflowProjectUrl,
+      'session-token': tokenResult.SessionToken,
+      'secret-access-key': tokenResult.SecretAccessKey,
+      'access-key-id': tokenResult.AccessKeyId,
+      expiration: tokenResult.Expiration,
+      region: tokenResult.Region,
+      'read-only': tokenResult.ReadOnly
+    }
+  };
+
+  let use = readOnly ? 'ReadOnly Access' : 'ReadWrite Access';
+
+  if (tokenUse) {
+    use = tokenUse;
+  }
+  const products = await prisma.products.findMany({
+    where: { ProjectId: projectId },
+    select: { Id: true }
+  });
+
+  await DatabaseWrites.productTransitions.createMany({
+    data: products.map((p) => ({
+      ProductId: p.Id,
+      AllowedUserNames: user[0].Name,
+      TransitionType: ProductTransitionType.ProjectAccess,
+      InitialState: 'Project ' + use,
+      WorkflowUserId: user[0].WorkflowUserId,
```

Comment:
Populate with user.Id as a string. Don't assign AllowedUserName. Display name base on WorkflowUserId.  If it is a GUID, look up user.WorkflowUserId. If it is just an int, lookup user.Id

Reply:
Alternatively, add ProductTransitions.UserId and S2 will populate that.